File: clients/csharp/WatchNexusClient.cs
// WatchNexus Licensing Server - drop-in .NET / C# client.
//
// Target: .NET 6.0+ (uses System.Net.Http.Json + System.Text.Json).
// Works in: console apps, WinForms/WPF/MAUI, ASP.NET, Unity (with HttpClient),
//           and any other .NET host with networking.
//
// Usage:
//
//     using WatchNexus;
//
//     var client = new WatchNexusClient(
//         baseUrl: "https://licenses.example.com",
//         apiKey:  "wnk_...",
//         licenseKey: "WNX-...");
//
//     var token = await client.ActivateAsync(new ActivateRequest {
//         HardwareId = "01:23:45:67:89:AB",
//         Domain     = "customer.example.com",
//         DeviceName = "Marie\u2019s Surface",
//     });
//
//     var state = await client.ValidateAsync(token);
//     // state.Valid + state.Mode ("online" | "grace" | "grace_offline" | ...)
//
//     await client.DeactivateAsync(token);
//
// Includes automatic offline-grace fallback in ValidateAsync (returns
// mode="grace_offline" when the server is unreachable BUT the locally
// decoded token is still within its grace_until window).
//
// MIT-licensed. Copy this single file into your project. No NuGet packages
// beyond what ships with .NET 6+.

using System;
using System.Net.Http;
using System.Net.Http.Json;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;

namespace WatchNexus;

public sealed class WatchNexusException : Exception
{
    public int? StatusCode { get; init; }
    public string? RawBody { get; init; }

    public WatchNexusException(string message, int? statusCode = null,
                               string? rawBody = null, Exception? inner = null)
        : base(message, inner)
    {
        StatusCode = statusCode;
        RawBody = rawBody;
    }
}

public record ActivateRequest
{
    [JsonPropertyName("license_key")]   public string? LicenseKey { get; init; }
    [JsonPropertyName("hardware_id")]   public string? HardwareId { get; init; }
    [JsonPropertyName("domain")]        public string? Domain     { get; init; }
    [JsonPropertyName("device_name")]   public string? DeviceName { get; init; }
}

public record ActivateResponse
{
    [JsonPropertyName("activation_id")]    public string ActivationId    { get; init; } = "";
    [JsonPropertyName("activation_token")] public string ActivationToken { get; init; } = "";
    [JsonPropertyName("expires_at")]       public long ExpiresAt         { get; init; }
    [JsonPropertyName("grace_until")]      public long GraceUntil        { get; init; }
    [JsonPropertyName("reused")]           public bool Reused            { get; init; }
}

public record ValidateResponse
{
    [JsonPropertyName("valid")]       public bool Valid { get; init; }
    [JsonPropertyName("mode")]        public string? Mode { get; init; }
    [JsonPropertyName("expires_at")]  public long? ExpiresAt  { get; init; }
    [JsonPropertyName("grace_until")] public long? GraceUntil { get; init; }
}

public sealed class WatchNexusClient : IDisposable
{
    private readonly HttpClient _http;
    private readonly bool _ownsClient;
    private readonly string _baseUrl;
    private readonly string _apiKey;
    private readonly string? _licenseKey;
    private readonly JsonSerializerOptions _json = new()
    {
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        PropertyNameCaseInsensitive = true,
    };

    public WatchNexusClient(string baseUrl, string apiKey, string? licenseKey = null,
                            HttpClient? httpClient = null,
                            TimeSpan? timeout = null)
    {
        if (string.IsNullOrWhiteSpace(baseUrl)) throw new ArgumentException("baseUrl is required");
        if (string.IsNullOrWhiteSpace(apiKey))  throw new ArgumentException("apiKey is required");
        _baseUrl = baseUrl.TrimEnd('/');
        _apiKey = apiKey;
        _licenseKey = licenseKey;
        if (httpClient is null)
        {
            _http = new HttpClient();
            _ownsClient = true;
        }
        else
        {
            _http = httpClient;
            _ownsClient = false;
        }
        _http.Timeout = timeout ?? TimeSpan.FromSeconds(10);
        _http.DefaultRequestHeaders.UserAgent.ParseAdd("watchnexus-dotnet/1.0");
        _http.DefaultRequestHeaders.Remove("X-API-Key");
        _http.DefaultRequestHeaders.Add("X-API-Key", _apiKey);
    }

    public async Task<ActivateResponse> ActivateAsync(ActivateRequest req,
                                                       CancellationToken ct = default)
    {
        var body = req with { LicenseKey = req.LicenseKey ?? _licenseKey };
        if (string.IsNullOrEmpty(body.LicenseKey))
            throw new ArgumentException("LicenseKey required (pass it or set licenseKey ctor arg)");
        return await PostAsync<ActivateResponse>("/api/integrate/activate", body, ct);
    }

    /// <summary>
    /// Validate the activation token against the server. If the call fails
    /// AND <paramref name="allowOfflineGrace"/> is true AND the local JWT is
    /// still within its grace_until, returns a synthetic
    /// <see cref="ValidateResponse"/> with <c>Mode = "grace_offline"</c>.
    /// </summary>
    public async Task<ValidateResponse> ValidateAsync(string activationToken,
        string? hardwareId = null, string? domain = null,
        bool allowOfflineGrace = true, CancellationToken ct = default)
    {
        var body = new { activation_token = activationToken, hardware_id = hardwareId, domain };
        try
        {
            return await PostAsync<ValidateResponse>("/api/integrate/validate", body, ct);
        }
        catch (WatchNexusException) when (allowOfflineGrace) { return OfflineFallback(activationToken); }
        catch (HttpRequestException) when (allowOfflineGrace) { return OfflineFallback(activationToken); }
        catch (TaskCanceledException) when (allowOfflineGrace) { return OfflineFallback(activationToken); }
    }

    public Task<ValidateResponse> ValidateAsync(ActivateResponse token,
        string? hardwareId = null, string? domain = null,
        bool allowOfflineGrace = true, CancellationToken ct = default)
        => ValidateAsync(token.ActivationToken, hardwareId, domain, allowOfflineGrace, ct);

    public async Task<bool> DeactivateAsync(string activationToken,
        CancellationToken ct = default)
    {
        var body = new { activation_token = activationToken, license_key = _licenseKey };
        var resp = await PostAsync<JsonElement>("/api/integrate/deactivate", body, ct);
        return resp.TryGetProperty("ok", out var ok) && ok.GetBoolean();
    }

    public Task<bool> DeactivateAsync(ActivateResponse token,
                                       CancellationToken ct = default)
        => DeactivateAsync(token.ActivationToken, ct);

    public async Task<string> GetPublicKeyPemAsync(CancellationToken ct = default)
    {
        var r = await _http.GetAsync($"{_baseUrl}/api/public-key", ct);
        r.EnsureSuccessStatusCode();
        var doc = await r.Content.ReadFromJsonAsync<JsonElement>(_json, ct);
        return doc.GetProperty("pem").GetString() ?? "";
    }

    // -------------------- helpers --------------------
    private async Task<T> PostAsync<T>(string path, object body, CancellationToken ct)
    {
        using var req = new HttpRequestMessage(HttpMethod.Post, $"{_baseUrl}{path}")
        {
            Content = JsonContent.Create(body, options: _json),
        };
        using var r = await _http.SendAsync(req, ct);
        var raw = await r.Content.ReadAsStringAsync(ct);
        if (!r.IsSuccessStatusCode)
        {
            throw new WatchNexusException(
                $"{path} failed: {(int)r.StatusCode}",
                statusCode: (int)r.StatusCode, rawBody: raw);
        }
        return JsonSerializer.Deserialize<T>(raw, _json)
               ?? throw new WatchNexusException($"{path}: empty body");
    }

    private static ValidateResponse OfflineFallback(string token)
    {
        var claims = DecodeJwtPayload(token);
        var now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
        if (claims is not null && claims.Value.TryGetProperty("grace_until", out var g)
            && g.GetInt64() >= now)
        {
            return new ValidateResponse
            {
                Valid = true,
                Mode = "grace_offline",
                ExpiresAt = claims.Value.TryGetProperty("exp", out var e) ? e.GetInt64() : null,
                GraceUntil = g.GetInt64(),
            };
        }
        return new ValidateResponse { Valid = false, Mode = "unreachable" };
    }

    /// <summary>Decode (without verifying) the JWT payload of an activation token.</summary>
    public static JsonElement? DecodeJwtPayload(string token)
    {
        try
        {
            var parts = token.Split('.');
            if (parts.Length != 3) return null;
            var payload = Base64UrlDecode(parts[1]);
            using var doc = JsonDocument.Parse(payload);
            return doc.RootElement.Clone();
        }
        catch { return null; }
    }

    private static byte[] Base64UrlDecode(string s)
    {
        s = s.Replace('-', '+').Replace('_', '/');
        switch (s.Length % 4) { case 2: s += "=="; break; case 3: s += "="; break; }
        return Convert.FromBase64String(s);
    }

    public void Dispose()
    {
        if (_ownsClient) _http.Dispose();
    }
}

File: clients/python/watchnexus_client.py
"""
WatchNexus Licensing Server - drop-in Python client.

Usage:
    from watchnexus_client import WatchNexusClient

    client = WatchNexusClient(
        base_url="https://licenses.example.com",
        api_key="wnk_...",
        license_key="WNX-...",
    )

    token = client.activate(hardware_id="01:23:45:67:89:AB",
                            domain="customer.example.com",
                            device_name="Marie’s MacBook Pro")
    # token is a dict with activation_token, expires_at, grace_until, ...

    state = client.validate(token, hardware_id=..., domain=...)
    # state is a dict { valid: bool, mode: str, license: ..., activation: ... }

    # When offline, client.validate falls through to local grace-period check
    # so brief network blips never lock a paying user out.

    client.deactivate(token)

Only stdlib + requests. MIT-licensed; copy it into your codebase.
"""
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import time
from dataclasses import dataclass, field
from typing import Any, Optional

try:
    import requests  # type: ignore
except ImportError as exc:  # pragma: no cover
    raise ImportError(
        "watchnexus_client requires `requests`. Install with: pip install requests"
    ) from exc


class WatchNexusError(Exception):
    """Raised for non-2xx responses from the licensing server."""

    def __init__(self, message: str, status: Optional[int] = None,
                 payload: Any = None) -> None:
        super().__init__(message)
        self.status = status
        self.payload = payload


@dataclass
class WatchNexusClient:
    base_url: str
    api_key: str
    license_key: Optional[str] = None
    timeout: float = 10.0
    session: requests.Session = field(default_factory=requests.Session)

    def __post_init__(self) -> None:
        self.base_url = self.base_url.rstrip("/")
        self.session.headers.update({
            "X-API-Key": self.api_key,
            "Content-Type": "application/json",
            "User-Agent": "watchnexus-python/1.0",
        })

    # -------------------- public API --------------------
    def activate(self, hardware_id: Optional[str] = None,
                  domain: Optional[str] = None,
                  device_name: Optional[str] = None,
                  license_key: Optional[str] = None) -> dict:
        key = license_key or self.license_key
        if not key:
            raise ValueError("license_key required (pass it or set client.license_key)")
        body = {"license_key": key, "hardware_id": hardware_id,
                "domain": domain, "device_name": device_name}
        return self._post("/api/integrate/activate", body)

    def validate(self, token: dict | str, hardware_id: Optional[str] = None,
                  domain: Optional[str] = None,
                  allow_offline_grace: bool = True) -> dict:
        """Validate against the server. If the network call fails AND the local
        token is still within its `grace_until`, returns { valid: True, mode: 'grace_offline' }."""
        activation_token = token["activation_token"] if isinstance(token, dict) else token
        body = {"activation_token": activation_token,
                "hardware_id": hardware_id, "domain": domain}
        try:
            return self._post("/api/integrate/validate", body)
        except (requests.RequestException, WatchNexusError) as e:
            if allow_offline_grace:
                local = decode_activation_token_locally(activation_token)
                if local and local.get("grace_until", 0) >= int(time.time()):
                    return {"valid": True, "mode": "grace_offline",
                            "claims": local, "error": str(e)}
            raise

    def deactivate(self, token: dict | str,
                    hardware_id: Optional[str] = None,
                    domain: Optional[str] = None) -> dict:
        if isinstance(token, dict):
            activation_token = token.get("activation_token")
        else:
            activation_token = token
        body = {"activation_token": activation_token,
                "license_key": self.license_key,
                "hardware_id": hardware_id, "domain": domain}
        return self._post("/api/integrate/deactivate", body)

    def public_key(self) -> str:
        r = self.session.get(f"{self.base_url}/api/public-key", timeout=self.timeout)
        r.raise_for_status()
        return r.json()["pem"]

    def health(self) -> dict:
        r = self.session.get(f"{self.base_url}/api/health", timeout=self.timeout)
        r.raise_for_status()
        return r.json()

    # -------------------- internals --------------------
    def _post(self, path: str, body: dict) -> dict:
        r = self.session.post(f"{self.base_url}{path}",
                              data=json.dumps(body), timeout=self.timeout)
        if r.status_code >= 400:
            try:
                payload = r.json()
            except Exception:
                payload = r.text
            raise WatchNexusError(
                f"{path} failed: {r.status_code}", status=r.status_code, payload=payload,
            )
        return r.json()


# ---------------------------------------------------------------------------
# Local helpers (no network) for offline grace handling.
# ---------------------------------------------------------------------------

def decode_activation_token_locally(token: str) -> Optional[dict]:
    """Decode (but do NOT cryptographically verify) the JWT payload so a
    client can fall back to grace-period checks while offline. Always
    pair this with a successful prior /activate against the server."""
    try:
        parts = token.split(".")
        if len(parts) != 3:
            return None
        pad = "=" * (-len(parts[1]) % 4)
        payload = base64.urlsafe_b64decode(parts[1] + pad)
        return json.loads(payload)
    except Exception:
        return None


def verify_license_key_hmac(license_key: str, shared_secret: bytes) -> Optional[dict]:
    """For HMAC-signed licenses, verify the key locally with the same secret
    your server uses (advanced; only useful if you embed the secret in a
    trusted environment)."""
    try:
        prefix, payload_b32, sig_b32 = license_key.split("-", 2)
        if prefix != "WNX":
            return None
        raw = _b32decode(payload_b32)
        sig = _b32decode(sig_b32)
        expected = hmac.new(shared_secret, raw, hashlib.sha256).digest()
        if not hmac.compare_digest(sig, expected):
            return None
        return json.loads(raw)
    except Exception:
        return None


def _b32decode(s: str) -> bytes:
    pad = "=" * (-len(s) % 8)
    return base64.b32decode(s + pad)
